include/dr/mhp/containers/distributed_dense_matrix.hpp

```diff
@@ -0,0 +1,323 @@
+    dm_rows_.reserve(segment_shape_[0]);
+
+    int row_start_index_ = 0;
+    for (auto _titr = rng::begin(segments_); _titr != rng::end(segments_);
+         ++_titr) {
+      for (int _ind = row_start_index_;
+           _ind < row_start_index_ + (int)(*_titr).shape()[0]; _ind++) {
+        T *_dataptr = nullptr;
+        if ((*_titr).is_local()) {
+          if (local_rows_indices_.first == -1)
+            local_rows_indices_.first = _ind;
+          local_rows_indices_.second = _ind;
+
+          int _dataoff = halo_bounds_.prev * shape_[1]; // start of data
+          _dataoff +=
+              (_ind - default_comm().rank() * segment_shape_[0]) * shape_[1];
+
+          assert(_dataoff >= 0);
+          assert(_dataoff < (int)data_size_);
+          _dataptr = data_ + _dataoff;
+        }
+        dm_rows_.emplace_back(_ind, _dataptr, shape_[1], &(*_titr));
+      }
+      row_start_index_ += (*_titr).shape()[0];
+    };
+
+    // rows in halo.prev area
+    for (int _ind = local_rows_indices_.first - halo_bounds_.prev;
+         _ind < local_rows_indices_.first; _ind++) {
+      std::size_t _dataoff =
+          (_ind + halo_bounds_.prev - local_rows_indices_.first) * shape_[1];
+
+      assert(_dataoff >= 0);
+      assert(_dataoff < halo_bounds_.prev * shape_[1]);
+      dm_halo_p_rows_.emplace_back(_ind, data_ + _dataoff, shape_[1],
+                                   &(*rng::begin(segments_)));
+    }
+
+    // rows in halo.next area
+    for (int _ind = local_rows_indices_.second + 1;
+         _ind < (int)(local_rows_indices_.second + 1 + halo_bounds_.next);
+         _ind++) {
+      int _dataoff =
+          (_ind + halo_bounds_.prev - local_rows_indices_.first) * shape_[1];
+
+      assert(_dataoff >= 0);
+      assert(_dataoff < (int)data_size_);
+      dm_halo_n_rows_.emplace_back(
+          _ind, data_ + _dataoff, shape_[1],
+          &(*(rng::begin(segments_) + default_comm().rank())));
+    }
+
+    win_.create(default_comm(), data_, data_size_ * sizeof(T));
+    active_wins().insert(win_.mpi_win());
+    dm_segments_ = dm_segments<distributed_dense_matrix>(this);
+    fence();
+  }
+
+private:
+  friend d_segment_iterator<distributed_dense_matrix>;
+  friend dm_segments<distributed_dense_matrix>;
+  friend dm_rows<distributed_dense_matrix>;
+  friend dm_rows_iterator<distributed_dense_matrix>;
+
+  key_type shape_;
+  std::size_t grid_size_; // distribution of tiles, currently (N, 1), and 2nd
+                          // dimention is omitted
+  key_type segment_shape_;
+
+  const std::size_t segment_size_ = 0; // size of local data
+  const std::size_t data_size_ = 0;    // all data with halo buffers
+
+  T *data_ = nullptr; // local data ptr
+
+  dr::span_halo<T> *halo_;
+  dr::halo_bounds halo_bounds_;
+
+  std::vector<d_segment<distributed_dense_matrix>> segments_;
+  dm_segments<distributed_dense_matrix>
+      dm_segments_; // lightweight view on segments_
+
+  dm_rows<distributed_dense_matrix<T>>
```

Comment:
one time you use `distributed_dense_matrix` and another time `distributed_dense_matrix<T>`
Stick to one notation. The one which is better if one of them is better than other.

Reply:
Shorter is better, fixed